Bottom: c26d863c5dc11498a04a7eec07edaff64c2bbe82
Top:    4f0766a71353baed0d4745709c899e5da27479e5
Author: Jason Wang <jasowang@redhat.com>
Date:   2016-11-24 11:09:10 +0800

tuntap: bypass unnecessary hash calculating

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index 64e694c..c31c9ca 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1294,7 +1294,8 @@ static ssize_t tun_get_user(struct tun_struct *tun, struct tun_file *tfile,
 	skb_reset_network_header(skb);
 	skb_probe_transport_header(skb, 0);
 
-	rxhash = skb_get_hash(skb);
+	if (tun->numqueues != 1 || static_key_false(&rps_needed))
+		rxhash = skb_get_hash(skb);
 	netif_rx_ni(skb);
 
 	stats = get_cpu_ptr(tun->pcpu_stats);
@@ -1304,7 +1305,8 @@ static ssize_t tun_get_user(struct tun_struct *tun, struct tun_file *tfile,
 	u64_stats_update_end(&stats->syncp);
 	put_cpu_ptr(stats);
 
-	tun_flow_update(tun, rxhash, tfile);
+	if (tun->numqueues != 1 || static_key_false(&rps_needed))
+		tun_flow_update(tun, rxhash, tfile);
 	return total_len;
 }
